File: sdk/php/tests/e2e/rest/Account/AccountSubAccountTest.php
<?php

namespace Tests\e2e\rest\Account;

use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Extension\Interceptor\Logging;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\AddSubAccountApiReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\AddSubAccountFuturesPermissionReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\AddSubAccountMarginPermissionReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\AddSubAccountReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\DeleteSubAccountApiReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\GetFuturesSubAccountListV2Req;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\GetSpotSubAccountDetailReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\GetSpotSubAccountListV2Req;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\GetSpotSubAccountsSummaryV2Req;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\GetSubAccountApiListReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\ModifySubAccountApiReq;
use KuCoin\UniversalSDK\Generate\Account\Subaccount\SubAccountApi;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use PHPUnit\Framework\TestCase;

class AccountSubAccountTest extends TestCase
{
    /**
     * @var SubAccountApi $api
     */
    private $api;

    /**
     * @var Serializer $serializer
     */
    private $serializer;

    protected function setUp(): void
    {

        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(
                    new JsonSerializedHandler()
                );
            })
            ->build();

        // Retrieve API secret information from environment variables
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        // Optional: Retrieve broker secret information from environment variables; applicable for broker API only
        $brokerName = getenv('BROKER_NAME');
        $brokerKey = getenv('BROKER_KEY');
        $brokerPartner = getenv('BROKER_PARTNER');

        // Set specific options, others will fall back to default values
        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setMaxConnections(10)
            ->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setBrokerName($brokerName)
            ->setBrokerKey($brokerKey)
            ->setBrokerPartner($brokerPartner)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->build();

        $client = new DefaultClient($clientOption);
        $kucoinRestService = $client->restService();
        $this->api = $kucoinRestService->getAccountService()->getSubAccountApi();
    }


    /**
     * addSubAccount
     * Add sub-account
     * /api/v2/sub/user/created
     */
    public function testAddSubAccount()
    {
        $builder = AddSubAccountReq::builder();
        $builder->setPassword("***@123")->setRemarks("****@123")->setSubName("sdkTestN")->setAccess("Spot");
        $req = $builder->build();
        $resp = $this->api->addSubAccount($req);
        self::assertNotNull($resp->uid);
        self::assertNotNull($resp->subName);
        self::assertNotNull($resp->remarks);
        self::assertNotNull($resp->access);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * addSubAccountMarginPermission
     * Add sub-account Margin Permission
     * /api/v3/sub/user/margin/enable
     */
    public function testAddSubAccountMarginPermission()
    {
        $builder = AddSubAccountMarginPermissionReq::builder();
        $builder->setUid("****");
        $req = $builder->build();
        $resp = $this->api->addSubAccountMarginPermission($req);
        self::assertNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * addSubAccountFuturesPermission
     * Add sub-account Futures Permission
     * /api/v3/sub/user/futures/enable
     */
    public function testAddSubAccountFuturesPermission()
    {
        $builder = AddSubAccountFuturesPermissionReq::builder();
        $builder->setUid("*****");
        $req = $builder->build();
        $resp = $this->api->addSubAccountFuturesPermission($req);
        self::assertNotNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSpotSubAccountsSummaryV2
     * Get sub-account List - Summary Info
     * /api/v2/sub/user
     */
    public function testGetSpotSubAccountsSummaryV2()
    {
        $builder = GetSpotSubAccountsSummaryV2Req::builder();
        $builder->setCurrentPage(1)->setPageSize(10);
        $req = $builder->build();
        $resp = $this->api->getSpotSubAccountsSummaryV2($req);
        self::assertNotNull($resp->currentPage);
        self::assertNotNull($resp->pageSize);
        self::assertNotNull($resp->totalNum);
        self::assertNotNull($resp->totalPage);
        foreach ($resp->items as $item) {
            self::assertNotNull($item->userId);
            self::assertNotNull($item->uid);
            self::assertNotNull($item->subName);
            self::assertNotNull($item->status);
            self::assertNotNull($item->type);
            self::assertNotNull($item->access);
            self::assertNotNull($item->createdAt);
            self::assertNotNull($item->remarks);
            self::assertNotNull($item->tradeTypes);
            self::assertNotNull($item->openedTradeTypes);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSpotSubAccountDetail
     * Get sub-account Detail - Balance
     * /api/v1/sub-accounts/{subUserId}
     */
    public function testGetSpotSubAccountDetail()
    {
        $builder = GetSpotSubAccountDetailReq::builder();
        $builder->setSubUserId("*****")->setIncludeBaseAmount(false)->
        setBaseCurrency("USDT")->setBaseAmount("0.1");
        $req = $builder->build();
        $resp = $this->api->getSpotSubAccountDetail($req);
        self::assertNotNull($resp->subUserId);
        self::assertNotNull($resp->subName);
        foreach ($resp->mainAccounts as $item) {
            self::assertNotNull($item->currency);
            self::assertNotNull($item->balance);
            self::assertNotNull($item->available);
            self::assertNotNull($item->holds);
            self::assertNotNull($item->baseCurrency);
            self::assertNotNull($item->baseCurrencyPrice);
            self::assertNotNull($item->baseAmount);
            self::assertNotNull($item->tag);
        }

        foreach ($resp->tradeAccounts as $item) {
            self::assertNotNull($item->currency);
            self::assertNotNull($item->balance);
            self::assertNotNull($item->available);
            self::assertNotNull($item->holds);
            self::assertNotNull($item->baseCurrency);
            self::assertNotNull($item->baseCurrencyPrice);
            self::assertNotNull($item->baseAmount);
            self::assertNotNull($item->tag);
        }

        foreach ($resp->marginAccounts as $item) {
            self::assertNotNull($item->currency);
            self::assertNotNull($item->balance);
            self::assertNotNull($item->available);
            self::assertNotNull($item->holds);
            self::assertNotNull($item->baseCurrency);
            self::assertNotNull($item->baseCurrencyPrice);
            self::assertNotNull($item->baseAmount);
            self::assertNotNull($item->tag);
        }

        foreach ($resp->tradeHFAccounts as $item) {
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSpotSubAccountListV2
     * Get sub-account List - Spot Balance (V2)
     * /api/v2/sub-accounts
     */
    public function testGetSpotSubAccountListV2()
    {
        $builder = GetSpotSubAccountListV2Req::builder();
        $builder->setCurrentPage(1)->setPageSize(10);
        $req = $builder->build();
        $resp = $this->api->getSpotSubAccountListV2($req);
        self::assertNotNull($resp->currentPage);
        self::assertNotNull($resp->pageSize);
        self::assertNotNull($resp->totalNum);
        self::assertNotNull($resp->totalPage);
        foreach ($resp->items as $item) {
            self::assertNotNull($item->subUserId);
            self::assertNotNull($item->subName);
            self::assertNotNull($item->mainAccounts);
            self::assertNotNull($item->tradeAccounts);
            self::assertNotNull($item->marginAccounts);
            self::assertNotNull($item->tradeHFAccounts);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getFuturesSubAccountListV2
     * Get sub-account List - Futures Balance (V2)
     * /api/v1/account-overview-all
     */
    public function testGetFuturesSubAccountListV2()
    {
        $builder = GetFuturesSubAccountListV2Req::builder();
        $builder->setCurrency("XBT");
        $req = $builder->build();
        $resp = $this->api->getFuturesSubAccountListV2($req);
        self::assertNotNull($resp->summary);
        foreach ($resp->accounts as $item) {
            self::assertNotNull($item->accountName);
            self::assertNotNull($item->accountEquity);
            self::assertNotNull($item->unrealisedPNL);
            self::assertNotNull($item->marginBalance);
            self::assertNotNull($item->positionMargin);
            self::assertNotNull($item->orderMargin);
            self::assertNotNull($item->frozenFunds);
            self::assertNotNull($item->availableBalance);
            self::assertNotNull($item->currency);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * addSubAccountApi
     * Add sub-account API
     * /api/v1/sub/api-key
     */
    public function testAddSubAccountApi()
    {
        $builder = AddSubAccountApiReq::builder();
        $builder->setPassphrase("********")->setRemark("remark3")->
        setIpWhitelist("192.1.1.1")->setPermission("General,Spot")->setExpire('30')->setSubName("sdkTestN");
        $req = $builder->build();
        $resp = $this->api->addSubAccountApi($req);
        self::assertNotNull($resp->subName);
        self::assertNotNull($resp->remark);
        self::assertNotNull($resp->apiKey);
        self::assertNotNull($resp->apiSecret);
        self::assertNotNull($resp->apiVersion);
        self::assertNotNull($resp->passphrase);
        self::assertNotNull($resp->permission);
        self::assertNotNull($resp->ipWhitelist);
        self::assertNotNull($resp->createdAt);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * modifySubAccountApi
     * Modify sub-account API
     * /api/v1/sub/api-key/update
     */
    public function testModifySubAccountApi()
    {
        $builder = ModifySubAccountApiReq::builder();
        $builder->setPassphrase("****")->setApiKey("******")->
        setIpWhitelist("192.1.1.2")->setPermission("General,Spot")->setExpire('30')->setSubName("sdkTestN");
        $req = $builder->build();
        $resp = $this->api->modifySubAccountApi($req);
        self::assertNotNull($resp->subName);
        self::assertNotNull($resp->apiKey);
        self::assertNotNull($resp->permission);
        self::assertNotNull($resp->ipWhitelist);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSubAccountApiList
     * Get sub-account API List
     * /api/v1/sub/api-key
     */
    public function testGetSubAccountApiList()
    {
        $builder = GetSubAccountApiListReq::builder();
        $builder->setApiKey("*******")->setSubName("sdkTestN");
        $req = $builder->build();
        $resp = $this->api->getSubAccountApiList($req);
        foreach ($resp->data as $item) {
            self::assertNotNull($item->subName);
            self::assertNotNull($item->remark);
            self::assertNotNull($item->apiKey);
            self::assertNotNull($item->apiVersion);
            self::assertNotNull($item->permission);
            self::assertNotNull($item->ipWhitelist);
            self::assertNotNull($item->createdAt);
            self::assertNotNull($item->uid);
            self::assertNotNull($item->isMaster);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * deleteSubAccountApi
     * Delete sub-account API
     * /api/v1/sub/api-key
     */
    public function testDeleteSubAccountApi()
    {
        $builder = DeleteSubAccountApiReq::builder();
        $builder->setApiKey("*****")->setSubName("sdkTestN")->setPassphrase("****");
        $req = $builder->build();
        $resp = $this->api->deleteSubAccountApi($req);
        self::assertNotNull($resp->subName);
        self::assertNotNull($resp->apiKey);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSpotSubAccountsSummaryV1
     * Get sub-account List - Summary Info (V1)
     * /api/v1/sub/user
     */
    public function testGetSpotSubAccountsSummaryV1()
    {
        $resp = $this->api->getSpotSubAccountsSummaryV1();
        foreach ($resp->data as $item) {
            self::assertNotNull($item->userId);
            self::assertNotNull($item->uid);
            self::assertNotNull($item->subName);
            self::assertNotNull($item->type);
            self::assertNotNull($item->remarks);
            self::assertNotNull($item->access);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getSpotSubAccountListV1
     * Get sub-account List - Spot Balance (V1)
     * /api/v1/sub-accounts
     */
    public function testGetSpotSubAccountListV1()
    {
        $resp = $this->api->getSpotSubAccountListV1();
        foreach ($resp->data as $item) {
            self::assertNotNull($item->subUserId);
            self::assertNotNull($item->subName);
            self::assertNotNull($item->mainAccounts);
            self::assertNotNull($item->tradeAccounts);
            self::assertNotNull($item->marginAccounts);
            self::assertNotNull($item->tradeHFAccounts);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }


}

File: sdk/java/src/test/java/com/kucoin/universal/sdk/test/e2e/rest/account/AccountSubAccountTest.java
package com.kucoin.universal.sdk.test.e2e.rest.account;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.generate.account.fee.*;
import com.kucoin.universal.sdk.generate.account.subaccount.*;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import java.io.IOException;
import java.util.Collections;
import lombok.extern.slf4j.Slf4j;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

@Slf4j
public class AccountSubAccountTest {

  private static SubAccountApi api;

  public static ObjectMapper mapper = new ObjectMapper();

  @BeforeAll
  public static void setUp() {

    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    TransportOption httpTransport =
        TransportOption.builder()
            .interceptors(
                Collections.singleton(
                    new Interceptor() {
                      @NotNull @Override
                      public Response intercept(@NotNull Chain chain) throws IOException {
                        Request request = chain.request();

                        System.out.println("========== Request ==========");
                        System.out.println(request.method() + " " + request.url());

                        Response response = chain.proceed(request);

                        System.out.println("========== Response ==========");
                        System.out.println("Status Code: " + response.code());
                        System.out.println("Message: " + response.message());
                        return response;
                      }
                    }))
            .build();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(httpTransport)
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    api = kucoinClient.getRestService().getAccountService().getSubAccountApi();
  }

  /** addSubAccount Add sub-account /api/v2/sub/user/created */
  @Test
  public void testAddSubAccount() throws Exception {
    AddSubAccountReq.AddSubAccountReqBuilder builder = AddSubAccountReq.builder();
    builder
        .password("********@123")
        .remarks("")
        .subName("sdkTestIntegration3")
        .access(AddSubAccountReq.AccessEnum.SPOT);
    AddSubAccountReq req = builder.build();
    AddSubAccountResp resp = api.addSubAccount(req);
    Assertions.assertNotNull(resp.getUid());
    Assertions.assertNotNull(resp.getSubName());
    Assertions.assertNotNull(resp.getRemarks());
    Assertions.assertNotNull(resp.getAccess());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * addSubAccountMarginPermission Add sub-account Margin Permission /api/v3/sub/user/margin/enable
   */
  @Test
  public void testAddSubAccountMarginPermission() throws Exception {
    AddSubAccountMarginPermissionReq.AddSubAccountMarginPermissionReqBuilder builder =
        AddSubAccountMarginPermissionReq.builder();
    builder.uid("248443791");
    AddSubAccountMarginPermissionReq req = builder.build();
    AddSubAccountMarginPermissionResp resp = api.addSubAccountMarginPermission(req);
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * addSubAccountFuturesPermission Add sub-account Futures Permission
   * /api/v3/sub/user/futures/enable
   */
  @Test
  public void testAddSubAccountFuturesPermission() throws Exception {
    AddSubAccountFuturesPermissionReq.AddSubAccountFuturesPermissionReqBuilder builder =
        AddSubAccountFuturesPermissionReq.builder();
    builder.uid("248443791");
    AddSubAccountFuturesPermissionReq req = builder.build();
    AddSubAccountFuturesPermissionResp resp = api.addSubAccountFuturesPermission(req);
    Assertions.assertNotNull(resp.getData());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSpotSubAccountsSummaryV2 Get sub-account List - Summary Info /api/v2/sub/user */
  @Test
  public void testGetSpotSubAccountsSummaryV2() throws Exception {
    GetSpotSubAccountsSummaryV2Req.GetSpotSubAccountsSummaryV2ReqBuilder builder =
        GetSpotSubAccountsSummaryV2Req.builder();
    builder.currentPage(1).pageSize(10);
    GetSpotSubAccountsSummaryV2Req req = builder.build();
    GetSpotSubAccountsSummaryV2Resp resp = api.getSpotSubAccountsSummaryV2(req);
    Assertions.assertNotNull(resp.getCurrentPage());
    Assertions.assertNotNull(resp.getPageSize());
    Assertions.assertNotNull(resp.getTotalNum());
    Assertions.assertNotNull(resp.getTotalPage());
    resp.getItems()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getUserId());
              Assertions.assertNotNull(item.getUid());
              Assertions.assertNotNull(item.getSubName());
              Assertions.assertNotNull(item.getStatus());
              Assertions.assertNotNull(item.getType());
              Assertions.assertNotNull(item.getAccess());
              Assertions.assertNotNull(item.getCreatedAt());
              Assertions.assertNotNull(item.getRemarks());
              Assertions.assertNotNull(item.getTradeTypes());
              Assertions.assertNotNull(item.getOpenedTradeTypes());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSpotSubAccountDetail Get sub-account Detail - Balance /api/v1/sub-accounts/{subUserId} */
  @Test
  public void testGetSpotSubAccountDetail() throws Exception {
    GetSpotSubAccountDetailReq.GetSpotSubAccountDetailReqBuilder builder =
        GetSpotSubAccountDetailReq.builder();
    builder
        .subUserId("6745b7bc890ba20001911363")
        .includeBaseAmount(false)
        .baseCurrency("USDT")
        .baseAmount("0.1");
    GetSpotSubAccountDetailReq req = builder.build();
    GetSpotSubAccountDetailResp resp = api.getSpotSubAccountDetail(req);
    Assertions.assertNotNull(resp.getSubUserId());
    Assertions.assertNotNull(resp.getSubName());
    resp.getMainAccounts()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getCurrency());
              Assertions.assertNotNull(item.getBalance());
              Assertions.assertNotNull(item.getAvailable());
              Assertions.assertNotNull(item.getHolds());
              Assertions.assertNotNull(item.getBaseCurrency());
              Assertions.assertNotNull(item.getBaseCurrencyPrice());
              Assertions.assertNotNull(item.getBaseAmount());
              Assertions.assertNotNull(item.getTag());
            });

    resp.getTradeAccounts()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getCurrency());
              Assertions.assertNotNull(item.getBalance());
              Assertions.assertNotNull(item.getAvailable());
              Assertions.assertNotNull(item.getHolds());
              Assertions.assertNotNull(item.getBaseCurrency());
              Assertions.assertNotNull(item.getBaseCurrencyPrice());
              Assertions.assertNotNull(item.getBaseAmount());
              Assertions.assertNotNull(item.getTag());
            });

    resp.getMarginAccounts()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getCurrency());
              Assertions.assertNotNull(item.getBalance());
              Assertions.assertNotNull(item.getAvailable());
              Assertions.assertNotNull(item.getHolds());
              Assertions.assertNotNull(item.getBaseCurrency());
              Assertions.assertNotNull(item.getBaseCurrencyPrice());
              Assertions.assertNotNull(item.getBaseAmount());
              Assertions.assertNotNull(item.getTag());
            });

    resp.getTradeHFAccounts().forEach(item -> {});

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSpotSubAccountListV2 Get sub-account List - Spot Balance (V2) /api/v2/sub-accounts */
  @Test
  public void testGetSpotSubAccountListV2() throws Exception {
    GetSpotSubAccountListV2Req.GetSpotSubAccountListV2ReqBuilder builder =
        GetSpotSubAccountListV2Req.builder();
    builder.currentPage(1).pageSize(10);
    GetSpotSubAccountListV2Req req = builder.build();
    GetSpotSubAccountListV2Resp resp = api.getSpotSubAccountListV2(req);
    Assertions.assertNotNull(resp.getCurrentPage());
    Assertions.assertNotNull(resp.getPageSize());
    Assertions.assertNotNull(resp.getTotalNum());
    Assertions.assertNotNull(resp.getTotalPage());
    resp.getItems()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getSubUserId());
              Assertions.assertNotNull(item.getSubName());
              Assertions.assertNotNull(item.getMainAccounts());
              Assertions.assertNotNull(item.getTradeAccounts());
              Assertions.assertNotNull(item.getMarginAccounts());
              Assertions.assertNotNull(item.getTradeHFAccounts());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * getFuturesSubAccountListV2 Get sub-account List - Futures Balance (V2)
   * /api/v1/account-overview-all
   */
  @Test
  public void testGetFuturesSubAccountListV2() throws Exception {
    GetFuturesSubAccountListV2Req.GetFuturesSubAccountListV2ReqBuilder builder =
        GetFuturesSubAccountListV2Req.builder();
    builder.currency("XBT");
    GetFuturesSubAccountListV2Req req = builder.build();
    GetFuturesSubAccountListV2Resp resp = api.getFuturesSubAccountListV2(req);
    Assertions.assertNotNull(resp.getSummary());
    resp.getAccounts()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getAccountName());
              Assertions.assertNotNull(item.getAccountEquity());
              Assertions.assertNotNull(item.getUnrealisedPNL());
              Assertions.assertNotNull(item.getMarginBalance());
              Assertions.assertNotNull(item.getPositionMargin());
              Assertions.assertNotNull(item.getOrderMargin());
              Assertions.assertNotNull(item.getFrozenFunds());
              Assertions.assertNotNull(item.getAvailableBalance());
              Assertions.assertNotNull(item.getCurrency());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** addSubAccountApi Add sub-account API /api/v1/sub/api-key */
  @Test
  public void testAddSubAccountApi() throws Exception {
    AddSubAccountApiReq.AddSubAccountApiReqBuilder builder = AddSubAccountApiReq.builder();
    builder
        .passphrase("******")
        .remark("******")
        .permission("General,Spot")
        .ipWhitelist("")
        .expire(AddSubAccountApiReq.ExpireEnum._30)
        .subName("sdkTestIntegration3");
    AddSubAccountApiReq req = builder.build();
    AddSubAccountApiResp resp = api.addSubAccountApi(req);
    Assertions.assertNotNull(resp.getApiKey());
    Assertions.assertNotNull(resp.getApiSecret());
    Assertions.assertNotNull(resp.getApiVersion());
    Assertions.assertNotNull(resp.getPassphrase());
    Assertions.assertNotNull(resp.getPermission());
    Assertions.assertNotNull(resp.getCreatedAt());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** modifySubAccountApi Modify sub-account API /api/v1/sub/api-key/update */
  @Test
  public void testModifySubAccountApi() throws Exception {
    ModifySubAccountApiReq.ModifySubAccountApiReqBuilder builder = ModifySubAccountApiReq.builder();
    builder
        .passphrase("******")
        .expire(ModifySubAccountApiReq.ExpireEnum._90)
        .subName("sdkTestIntegration3")
        .apiKey("6880428d28335c0001f5b45d");
    ModifySubAccountApiReq req = builder.build();
    ModifySubAccountApiResp resp = api.modifySubAccountApi(req);
    Assertions.assertNotNull(resp.getSubName());
    Assertions.assertNotNull(resp.getApiKey());
    Assertions.assertNotNull(resp.getPermission());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSubAccountApiList Get sub-account API List /api/v1/sub/api-key */
  @Test
  public void testGetSubAccountApiList() throws Exception {
    GetSubAccountApiListReq.GetSubAccountApiListReqBuilder builder =
        GetSubAccountApiListReq.builder();
    builder.apiKey("6880428d28335c0001f5b45d").subName("sdkTestIntegration3");
    GetSubAccountApiListReq req = builder.build();
    GetSubAccountApiListResp resp = api.getSubAccountApiList(req);
    resp.getData()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getSubName());
              Assertions.assertNotNull(item.getRemark());
              Assertions.assertNotNull(item.getApiKey());
              Assertions.assertNotNull(item.getApiVersion());
              Assertions.assertNotNull(item.getPermission());
              Assertions.assertNotNull(item.getCreatedAt());
              Assertions.assertNotNull(item.getUid());
              Assertions.assertNotNull(item.getIsMaster());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** deleteSubAccountApi Delete sub-account API /api/v1/sub/api-key */
  @Test
  public void testDeleteSubAccountApi() throws Exception {
    DeleteSubAccountApiReq.DeleteSubAccountApiReqBuilder builder = DeleteSubAccountApiReq.builder();
    builder
        .apiKey("6880428d28335c0001f5b45d")
        .subName("sdkTestIntegration3")
        .passphrase("123hhhhakjsbc");
    DeleteSubAccountApiReq req = builder.build();
    DeleteSubAccountApiResp resp = api.deleteSubAccountApi(req);
    Assertions.assertNotNull(resp.getSubName());
    Assertions.assertNotNull(resp.getApiKey());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSpotSubAccountsSummaryV1 Get sub-account List - Summary Info (V1) /api/v1/sub/user */
  @Test
  public void testGetSpotSubAccountsSummaryV1() throws Exception {
    GetSpotSubAccountsSummaryV1Resp resp = api.getSpotSubAccountsSummaryV1();
    resp.getData()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getUserId());
              Assertions.assertNotNull(item.getUid());
              Assertions.assertNotNull(item.getSubName());
              Assertions.assertNotNull(item.getType());
              Assertions.assertNotNull(item.getRemarks());
              Assertions.assertNotNull(item.getAccess());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** getSpotSubAccountListV1 Get sub-account List - Spot Balance (V1) /api/v1/sub-accounts */
  @Test
  public void testGetSpotSubAccountListV1() throws Exception {
    GetSpotSubAccountListV1Resp resp = api.getSpotSubAccountListV1();
    resp.getData()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getSubUserId());
              Assertions.assertNotNull(item.getSubName());
              Assertions.assertNotNull(item.getMainAccounts());
              Assertions.assertNotNull(item.getTradeAccounts());
              Assertions.assertNotNull(item.getMarginAccounts());
              Assertions.assertNotNull(item.getTradeHFAccounts());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }
}
